// pouchdb-errors-poc/src/shared/constants/errorStatus.tsx
const ErrorStatus = {
    UNAUTHORIZED: {
      status: 401,
      error: 'unauthorized',
      reason: 'Name or password is incorrect'
    },
    MISSING_BULK_DOCS: {
      status: 400,
      error: 'bad_request',
      reason: `Missing JSON list of 'docs'`
    },
    MISSING_DOC: {
      status: 404,
      error: 'not_found',
      reason: 'missing'
    },
    REV_CONFLICT: {
      status: 409,
      error: 'conflict',
      reason: 'Document update conflict'
    },
    INVALID_ID: {
      status: 400,
      error: 'bad_request',
      reason: '_id field must contain a string'
    },
    MISSING_ID: {
      status: 412,
      error: 'missing_id',
      reason: '_id is required for puts'
    },
    RESERVED_ID: {
      status: 400,
      error: 'bad_request',
      reason: 'Only reserved document ids may start with underscore'
    },
    NOT_OPEN: {
      status: 412,
      error: 'precondition_failed',
      reason: 'Database not open'
    },
    UNKNOWN_ERROR: {
      status: 500,
      error: 'unknown_error',
      reason: 'Database encountered an unknown error'
    },
    BAD_ARG: {
      status: 500,
      error: 'bad_arg',
      reason: 'Some query argument is invalid'
    },
    INVALID_REQUEST: {
      status: 400,
      error: 'invalid_request',
      reason: 'Request was invalid'
    },
    QUERY_PARSE_ERROR: {
      status: 400,
      error: 'query_parse_errors',
      reason: 'Some query parameter is invalid'
    },
    DOC_VALIDATION: {
      status: 500,
      error: 'doc_validation',
      reason: 'Bad special document member'
    },
    BAD_REQUEST: {
      status: 400,
      error: 'bad_request',
      reason: 'Something wrong with the request'
    },
    NOT_AN_OBJECT: {
      status: 400,
      error: 'bad_request',
      reason: 'Document must be a JSON object'
    },
    DB_MISSING: {
      status: 404,
      error: 'not_found',
      reason: 'Database not found'
    },
    IDB_ERROR: {
      status: 500,
      error: 'unknown',
      reason: 'indexed_db_went_bad'
    },
    WSQ_ERROR: {
      status: 500,
      error: 'web_sql_went_bad',
      reason: 'unknown'
    },
    LDB_ERROR: {
      status: 500,
      error: 'levelDB_went_went_bad',
      reason: 'unknown'
    },
    FORBIDDEN: {
      status: 403,
      error: 'forbidden',
      reason: 'Forbidden by design doc validate_doc_update function'
    },
    INVALID_REV: {
      status: 400,
      error: 'bad_request',
      reason: 'Invalid rev format'
    },
    FILE_EXISTS: {
      status: 412,
      error: 'file_exists',
      reason: 'The database could not be created, the file already exists'
    },
    MISSING_STUB: {
      status: 412,
      error: 'missing_stub',
      reason: 'A pre-existing attachment stub wasn\'t found'
    },
    INVALID_URL: {
      status: 413,
      error: 'invalid_url',
      reason: 'Provided URL is invalid'
    },
  }


export default ErrorStatus;
