// pouchdb-for-coverage/src/errors.js
import {
   UNAUTHORIZED,
   MISSING_BULK_DOCS,
   MISSING_DOC,
   REV_CONFLICT,
   INVALID_ID,
   MISSING_ID,
   RESERVED_ID,
   NOT_OPEN,
   UNKNOWN_ERROR,
   BAD_ARG,
   INVALID_REQUEST,
   QUERY_PARSE_ERROR,
   DOC_VALIDATION,
   BAD_REQUEST,
   NOT_AN_OBJECT,
   DB_MISSING,
   WSQ_ERROR,
   LDB_ERROR,
   FORBIDDEN,
   INVALID_REV,
   FILE_EXISTS,
   MISSING_STUB,
   IDB_ERROR,
   INVALID_URL
} from 'pouchdb-errors';

export default {
  UNAUTHORIZED: UNAUTHORIZED,
  MISSING_BULK_DOCS: MISSING_BULK_DOCS,
  MISSING_DOC: MISSING_DOC,
  REV_CONFLICT: REV_CONFLICT,
  INVALID_ID: INVALID_ID,
  MISSING_ID: MISSING_ID,
  RESERVED_ID: RESERVED_ID,
  NOT_OPEN: NOT_OPEN,
  UNKNOWN_ERROR: UNKNOWN_ERROR,
  BAD_ARG: BAD_ARG,
  INVALID_REQUEST: INVALID_REQUEST,
  QUERY_PARSE_ERROR: QUERY_PARSE_ERROR,
  DOC_VALIDATION: DOC_VALIDATION,
  BAD_REQUEST: BAD_REQUEST,
  NOT_AN_OBJECT: NOT_AN_OBJECT,
  DB_MISSING: DB_MISSING,
  WSQ_ERROR: WSQ_ERROR,
  LDB_ERROR: LDB_ERROR,
  FORBIDDEN: FORBIDDEN,
  INVALID_REV: INVALID_REV,
  FILE_EXISTS: FILE_EXISTS,
  MISSING_STUB: MISSING_STUB,
  IDB_ERROR: IDB_ERROR,
  INVALID_URL: INVALID_URL
};
