// pouchdb-find/src/adapters/http/index.js
import { createError, generateErrorFromResponse } from 'pouchdb-errors';
import { Headers } from 'pouchdb-fetch';
import massageCreateIndexRequest from '../../massageCreateIndexRequest';
import validateSelector from '../../validateSelector';

async function dbFetch(db, path, opts) {
  if (opts.body) {
    opts.body = JSON.stringify(opts.body);
    opts.headers = new Headers({ 'Content-type': 'application/json' });
  }

  const response = await db.fetch(path, opts);
  const json = await response.json();
  if (!response.ok) {
    json.status = response.status;
    throw generateErrorFromResponse(createError(json));
  }
  return json;
}

function createIndex(db, requestDef, callback) {
  dbFetch(db, '_index', {
    method: 'POST',
    body: massageCreateIndexRequest(requestDef)
  })
    .then((result) => {
      callback(null, result);
    })
    .catch(callback);
}

function find(db, requestDef, callback) {
  validateSelector(requestDef.selector, true);
  dbFetch(db, '_find', {
    method: 'POST',
    body: requestDef
  })
    .then((result) => {
      callback(null, result);
    })
    .catch(callback);
}

function explain(db, requestDef, callback) {
  dbFetch(db, '_explain', {
    method: 'POST',
    body: requestDef
  })
    .then((result) => {
      callback(null, result);
    })
    .catch(callback);
}

function getIndexes(db, callback) {
  return dbFetch(db, '_index', {
    method: 'GET',
  })
    .then((result) => {
      callback(null, result);
    })
    .catch(callback);
}

function deleteIndex(db, indexDef, callback) {
  const ddoc = indexDef.ddoc;
  const type = indexDef.type || 'json';
  const name = indexDef.name;

  if (!ddoc) {
    return callback(new Error('you must provide an index\'s ddoc'));
  }

  if (!name) {
    return callback(new Error('you must provide an index\'s name'));
  }

  const url = '_index/' + [ddoc, type, name].map(encodeURIComponent).join('/');

  dbFetch(db, url, { method: 'DELETE' })
    .then((result) => {
      callback(null, result);
    })
    .catch(callback);
}

export {
  createIndex,
  find,
  getIndexes,
  deleteIndex,
  explain
};
